segmentstore/server/src/main/java/io/pravega/segmentstore/server/attributes/ContainerAttributeIndexImpl.java

```diff
@@ -0,0 +1,152 @@
+    @GuardedBy("attributeIndices")
+    private final HashMap<Long, CompletableFuture<AttributeIndex>> attributeIndices;
+    private final ScheduledExecutorService executor;
+    private final String traceObjectId;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the ContainerAttributeIndexImpl class.
+     *
+     * @param containerMetadata The Segment Container's Metadata.
+     * @param storage           A Storage adapter which can be used to access the Attribute Segment.
+     * @param operationLog      An OperationLog that can be used to atomically update attributes for the main Segment.
+     * @param config            Attribute Index Configuration.
+     * @param executor          An Executor to run async tasks.
+     */
+    ContainerAttributeIndexImpl(ContainerMetadata containerMetadata, Storage storage, OperationLog operationLog,
+                                AttributeIndexConfig config, ScheduledExecutorService executor) {
+        this.containerMetadata = Preconditions.checkNotNull(containerMetadata, "containerMetadata");
+        this.storage = Preconditions.checkNotNull(storage, "storage");
+        this.operationLog = Preconditions.checkNotNull(operationLog, "operationLog");
+        this.config = Preconditions.checkNotNull(config, "config");
+        this.executor = Preconditions.checkNotNull(executor, "executor");
+        this.attributeIndices = new HashMap<>();
+        this.traceObjectId = String.format("ContainerAttributeIndex[%d]", containerMetadata.getContainerId());
+    }
+
+    //endregion
+
+    //region ContainerAttributeIndex Implementation
+
+    @Override
+    public CompletableFuture<AttributeIndex> forSegment(long streamSegmentId, Duration timeout) {
+        SegmentMetadata sm = this.containerMetadata.getStreamSegmentMetadata(streamSegmentId);
+        if (sm.isDeleted()) {
+            return Futures.failedFuture(new StreamSegmentNotExistsException(sm.getName()));
+        }
+
+        // Figure out if we already have this AttributeIndex cached. If not, we need to initialize it.
+        CompletableFuture<AttributeIndex> result;
+        AtomicReference<SegmentAttributeIndex> toInitialize = new AtomicReference<>();
+        synchronized (this.attributeIndices) {
+            result = this.attributeIndices.computeIfAbsent(streamSegmentId, id -> {
+                toInitialize.set(new SegmentAttributeIndex(sm, this.storage, this.operationLog, this.config, this.executor));
+                return new CompletableFuture<>();
+            });
+        }
+
+        if (toInitialize.get() == null) {
+            // We already have it cached - return its future (which should be already completed or will complete once
+            // its initialization is done).
+            return result;
+        } else {
+            try {
+                // Need to initialize the AttributeIndex and complete the future that we just registered.
+                // If this fails
```

Comment:
This comment line seems incomplete. Did you mean to talk about the call to `indexInitializationFailed`?

Reply:
Yes. Fixed.